src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotPrimaryDataQueryInterface.java

```diff
@@ -1,39 +1,77 @@
 import bio.terra.model.SnapshotRequestQueryModel;
 import bio.terra.service.common.CommonFlightUtils;
 import bio.terra.service.dataset.AssetSpecification;
 import bio.terra.service.dataset.Dataset;
 import bio.terra.service.dataset.DatasetService;
 import bio.terra.service.snapshot.Snapshot;
+import bio.terra.service.snapshot.SnapshotDao;
 import bio.terra.service.snapshot.exception.AssetNotFoundException;
+import bio.terra.service.snapshotbuilder.SnapshotBuilderService;
+import bio.terra.service.snapshotbuilder.SnapshotRequestDao;
 import bio.terra.stairway.FlightContext;
 import bio.terra.stairway.StepResult;
 import java.time.Instant;
 import java.util.List;
+import java.util.UUID;
 
 public interface CreateSnapshotPrimaryDataQueryInterface {
 
   default StepResult prepareQueryAndCreateSnapshot(
       FlightContext context,
       Snapshot snapshot,
       SnapshotRequestModel snapshotReq,
-      DatasetService datasetService)
+      DatasetService datasetService,
+      SnapshotBuilderService snapshotBuilderService,
+      SnapshotRequestDao snapshotRequestDao,
+      SnapshotDao snapshotDao,
+      AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+    AssetSpecification assetSpecification;
+    String sqlQuery;
+    Instant createdAt;
 
-    Query query = Query.parse(snapshotQuerySpec.getQuery());
-    String datasetName = query.getDatasetName();
-    Dataset dataset = datasetService.retrieveByName(datasetName);
-    AssetSpecification assetSpecification =
-        retrieveAssetSpecification(dataset, snapshotQuerySpec.getAssetName());
-    validateRootTable(query, assetSpecification);
-    String sqlQuery = translateQuery(query, dataset);
+    if (snapshotReq
+        .getContents()
+        .get(0)
+        .getMode()
+        .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
+      UUID accessRequestId =
+          snapshotReq.getContents().get(0).getRequestIdSpec().getSnapshotRequestId();
+      SnapshotAccessRequestResponse accessRequest = snapshotRequestDao.getById(accessRequestId);
+
+      UUID dataReleaseSnapshotId = accessRequest.getSourceSnapshotId();
+      Snapshot dataReleaseSnapshot = snapshotDao.retrieveSnapshot(dataReleaseSnapshotId);
+      // get the underlying dataset for the snapshot
+      if (dataReleaseSnapshot.getSnapshotSources().isEmpty()) {
+        throw new IllegalArgumentException("Snapshot does not have a source dataset");
+      }
+      Dataset dataset = dataReleaseSnapshot.getSnapshotSources().get(0).getDataset();
+      // gets pre-existing asset on the dataset
+      // TODO: create custom asset
+      assetSpecification = dataset.getAssetSpecificationByName("concept_asset").orElseThrow();
+      sqlQuery =
+          snapshotBuilderService.generateRowIdQuery(accessRequest, dataReleaseSnapshot, userReq);
+      createdAt = dataReleaseSnapshot.getCreatedDate();
```

Comment:
I'd like to talk through options for moving this code into its own step. It would be good if we could avoid having snapshot type logic embedded into the step. I was thinking we'd be able to convert the ByRequestId request into a ByQuery request so we wouldn't need to have special handling here.

Reply:
we could do that. We could have a step that either fully converts to the ByQuery model, or that converts just to an asset and a query. Then we would need to store it in the map and retrieve it here. So, there would still be special handling here either way (to retrieve from the map in the case of ByRequestId), but we could move the logic of creating the asset and query into another step.